Cache case id list

benrudolph opened the pull request:
@czue @wpride this will cache case id list when `cache_bust` is set to `false`. By default it will not cache so this won't break any api that doesn't want to cache. I used the skippable_quickcache which is very nice, but makes it hard for me to clear that key on a form submission. Basically I'd have to regenerate the key that's being created when this function is called which I don't think will be trivial and has room for a lot of errors (at least that was my thinking, happy to discuss).

Touchforms will know if it doesn't have an updated list of case ids because it will fail to find the case it's looking for. On failure it will tell HQ to bust the cache and try again.

This is functional, but I wanted to get it reviewed before I wrote a few tests for it

elf: @dannyroberts

Commit:
Cache case id list

benrudolph commented:
touchforms PR: https://github.com/dimagi/touchforms/pull/134

benrudolph commented on corehq/apps/cloudcare/views.py:
```diff
@@ -220,7 +221,26 @@ def form_context(request, domain, app_id, module_id, form_id):
+
 @cloudcare_api
+@skippable_quickcache(get_cases_vary_on, get_cases_skip_arg, timeout=50 * 60)
 def get_cases(request, domain):
```
I was thinking about this a bit, and I think the right approach might be to make another function called `_get_cases` that takes actual arguments (not request, and domain). That was I can use `skippable_quickcache` without modifying it and without using functions as the parameters. aaand maybe we might be able to reuse it for `filter_cases`

benrudolph replied:
actually maybe not, if i were to do that i should put the `skippable_quickcache` on the `get_filtered_cases` function. if i do that i don't think it'll work anymore because it'll cache the case list. it may be best to leave this where it is because it more clearly denotes the fact that all i want to do is cache the specific route

czue commented on corehq/apps/cloudcare/views.py:
```diff
@@ -220,7 +221,26 @@ def form_context(request, domain, app_id, module_id, form_id):
+
+
+def get_cases_skip_arg(request, domain):
+    return string_to_boolean(request.REQUEST.get('cache_bust', 'true'))
```
just to confirm, this only caches when you pass in `cache_bust=False`? I think that's right, I'm just curious how you came up with that convention (as opposed to e.g. `cache=True`)

czue replied:
Did you also only want to cache when `ids_only=True`? otherwise I'm not sure what would prevent this from caching the full results not just the ID lits.

benrudolph replied:
yea i initially didn't think it'd be a problem to cache the full result, but i only tested the simple case. as in: touchforms makes the call to grab the ids, then makes the call to grab the cases.  seems safer to just cache the ids. also you're totally right about the `cache_bust` thing. when i first wrote this, it was caching by default and i needed a param to break the cache. will update

Commit:
Modify API and add tests

Commit:
Quick cache comments and tests

benrudolph commented:
Updated with tests and addressed comments. @czue I decided to use the keyword `use_cache`. I was getting confused with the keyword `cache` since technically it's always caching the result it's whether or not we decide to skip the cache for the request.

The logic seems kind of brittle to me, but I'm hoping this will only be a short term fix

benrudolph commented on corehq/apps/cloudcare/tests/test_api.py:
```diff
@@ -10,9 +14,8 @@
     CASE_STATUS_CLOSED
-import uuid
 
-TEST_DOMAIN = "test-domain-uCSArw4SkbM6F04"
```
we should never use capital letters in domain names. it throws off validation since we normalize the domain names when we save them

benrudolph committed:
Update doc string

dimagimon commented on corehq/apps/cloudcare/views.py:
```diff
@@ -220,7 +221,35 @@ def form_context(request, domain, app_id, module_id, form_id):
+    When this function returns True, skippable_quickcache will not go to the cache for the result. By default,
+    if neither of these params are passed into the function, nothing will be cached. Cache will always be
+    skipped if ids_only is false.
+    
```
W293 blank line contains whitespace

benrudolph committed:
remove whitespace

czue commented:
:+1: to merge from my side pending possible QA

benrudolph commented:
One concern i thought of and not sure if we should be concerned. any time there's like a count on the number of cases in the form, and the cache case ids is cached, it will be wrong. situations like this, where there are calculations based on statistics of the case count will be stale (unless of course the case added was the case being viewed). do you think we should be OK with the staleness? or come up with a better cache busting technique? i think on submission would work, but we'd have to be more clever about how we are caching the case_id list.

with exception of that situation, i think this is g2g, but leaving the do not merge on until i confirm that the above situation would be acceptable

czue commented:
@benrudolph let's talk through this - grab me at some point today.

Commit:
Feature flag for case list cache

benrudolph commented:
Added a filter and checked to make sure nothing was using the cache if the flag wasn't set in the browser and tests. If you don't update the flag, the tests will fail. This was kind of a cute way to skip caching based on a feature flag. Let me know if this is too lazy though

czue commented:
good to merge when tests pass
